Subtitles refactoring, code cleanup and docstrings

ratoaq2 opened the pull request:
- Only subtitles.py was refactored in order to make the code more readable and maintainable
- Big methods were splittled into small methods
- Refactored duplicated code
- Added docstring to the methods
- Pre-scripts are now executed when post-processing is NOT postponed (before it was executed only when postponed; extra scripts are still not executed when postponed)
- Small fix in release refiner (missing . to separate file extension)

Do NOT merge it yet
Tests are "in progress"
My tests were all successful, but I want to give it more time and execute more scenarios.

OmgImAlexis commented on sickbeard/subtitles.py:
```diff
@@ -70,7 +84,8 @@ def sorted_service_list():
         if current_service in provider_manager.names():
             new_list.append({'name': current_service,
-                             'url': PROVIDER_URLS[current_service] if current_service in PROVIDER_URLS else lmgtfy % current_service,
+                             'url': PROVIDER_URLS[current_service]
```
This should really be like so to prevent long line length.

````
new_list.append({
    'name': current_service,
     'url': PROVIDER_URLS[current_service]
     if current_service in PROVIDER_URLS else lmgtfy % current_service,
     'image': current_service + '.png',
     'enabled': sickbeard.SUBTITLES_SERVICES_ENABLED[current_index] == 1
 })
````

fernandog commented:
@ratoaq2 how do you handle this removed code?

`language_exceptions = {'pt-br': 'pob'}`

ratoaq2 commented:
[Here](https://github.com/pymedusa/SickRage/blob/subtitles-cleanup/sickbeard/subtitles.py#L564)
`pt-BR` (or `pt-br`) is a valid IETF language code. Now it will support pt-BR and other IETF language codes as well and we do not need to maintain an exception list.
It will first try to detected it as an opensubtitles language code then it will fallback to IETF language code.

fernandog commented:
testing now

fernandog commented:
@ratoaq2 

```
2016-03-28 09:16:45 Thread-16 :: [f78217d] Failed doing webui callback: Traceback (most recent call last):
  File "/home/osmc/SickRage/sickbeard/webserve.py", line 304, in async_call
    result = function(**kwargs)
  File "/home/osmc/SickRage/sickbeard/webserve.py", line 3811, in index
    subtitlesFinderStatus=sickbeard.subtitlesFinderScheduler.action.amActive,
AttributeError: 'SubtitlesFinder' object has no attribute 'amActive'
```

ratoaq2 commented:
Thanks for reporting... That's fixed now.

fernandog commented:
Approved

fernandog commented:
@felipesalomao mudamos de fork. estamos usando esse. retiramos o limite de procurar por legendas. ainda nao temos um "release oficial" e portanto só os devs estão aqui e usando o "develop" branch. Se quiser testar. esteja a vontade. mas nao vamos dar muito suporte ativo

ratoaq2 commented:
This last commit enhances 2 things:

- Added cache for Provider Pool (it will prevent re-creating instances of the same provider when searching subtitles for multiple videos in a short period of time, as a consequence, it will reuse the same session, avoiding several logins with the same account for providers that needs authentication)
- Added cache for the video object (since the video file is static and its metadata will not change). This will reduce I/O and it will prevent HDD spin up when searching subtitles.

ratoaq2 commented:
Rebased/Squashed

Awaiting approval and it's ready to merge

fernandog commented on sickbeard/subtitles.py:
```diff
@@ -78,266 +96,546 @@ def sorted_service_list():
+        sickbeard.helpers.chmodAsParent(new_subtitles_path)
+    else:
+        logger.log(u'Unable to create subtitles folder {0}'.format(new_subtitles_path), logger.ERROR)
 
```
@ratoaq2  you cant set ERROR without traceback or exception. Change to WARNING

ratoaq2 replied:
Sure... I'll change it later today.
The [old code](https://github.com/pymedusa/SickRage/pull/219/files?diff=split#diff-425be5582fb3c7b0fce3dbcd319fed75L309) was ERROR as well:

fernandog commented on sickbeard/subtitles.py:
```diff
@@ -565,68 +798,93 @@ def dhm(td):
-                            episode_num(ep_to_sub['season'], ep_to_sub['episode'], numbering='absolute'), ex(error)), logger.ERROR)
+                logger.log(u'Error while searching subtitles for {0} {1}. Error: {2}'.format
+                           (ep_to_sub['show_name'], ep_num, ex(error)), logger.ERROR)
                 continue
```
@ratoaq2 here too. I know is old code. Can you change it please?

ratoaq2 replied:
Sure... I'll do it later. The [old code](https://github.com/pymedusa/SickRage/pull/219/files?diff=split#diff-425be5582fb3c7b0fce3dbcd319fed75L595) was error as well

fernandog commented:
@ratoaq2 im noticing a AA in logs:

This is usually happen when some error happens but there is no traceback

```
2016-03-29 07:39:05 INFO     FINDSUBTITLES :: [d9e372d] No subtitles found for /media/SAMSUNG/media/downloaded/series/Sl33py.S03E16.Dawns.Early.Light.1080p.Web-DL.DD5.1.H.264-DJMagic.mkv
2016-03-29 07:39:05 DEBUG    FINDSUBTITLES :: [d9e372d] Could not search in legendastv provider. Discarding for now
AA
2016-03-29 07:39:05 DEBUG    FINDSUBTITLES :: [d9e372d] Script result: Nothing to rename
```

ratoaq2 commented:
This message `Could not search` is present in only one place in the code -> [Here](https://github.com/pymedusa/SickRage/blob/subtitles-cleanup/sickbeard/subtitles.py#L313) after calling `pool.list_subtitles(video, languages)`

If any exception was thrown, it was handled by subliminal in that call. Is it possible to enable subliminal logging to inspect that issue?

fernandog commented:
Looks like the issue is when I 'print "Nothing to rename" in my script. Whenever I print something the AA appears

pre script: python tename_releases.py

```
# coding=utf-8
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import sys

filename = sys.argv[1]
if filename.rpartition('.')[2] in ('mkv', 'mp4', 'avi'):
    new_filename = filename.replace('VietHD', 'RARBG').replace('DRACULA','RARBG')
    if new_filename != filename:
        print "Renaming: {} para {}".format(filename, new_filename)
        os.rename(filename, new_filename)
    else:
        print "Nothing to rename"
```

fernandog commented on sickbeard/subtitles.py:
```diff
@@ -565,68 +798,93 @@ def dhm(td):
                                        stderr=subprocess.STDOUT, cwd=sickbeard.PROG_DIR)
             out, _ = process.communicate()  # @UnusedVariable
-            logger.log(u'Script result: {}'.format(out), logger.DEBUG)
+            logger.log(u'Script result: {0}'.format(out), logger.DEBUG)
```
@ratoaq2 this fixes the AA:

`logger.log(u'Script result: {0}'.format(out.strip()), logger.DEBUG)`

ratoaq2 replied:
:open_mouth:

ratoaq2 replied:
But `communicate() returns a tuple (stdout, stderr).` 
out is not a string... it's a tuple

labrys commented:
Approved!

fernandog commented:
https://github.com/pymedusa/SickRage/pull/617

fernandog commented on sickbeard/subtitles.py:
```diff
@@ -347,138 +645,58 @@ def __init__(self):
+                video_path = os.path.join(root, filename)
+                release_name = os.path.splitext(filename)[0]
+                found_subtitles = download_best_subs(video_path, root, release_name, languages, subtitles=False,
+                                                     embedded_subtitles=False, provider_pool=pool)
```
@ratoaq2 shouldn't be  `embedded_subtitles=bool(not sickbeard.EMBEDDED_SUBTITLES_ALL)`

so if video file has embedded subs and user don't want to ignore it, it can be PP with that sub only

as @NicoLeOca requested: can we get a checkbox "consider embedded subs as of default language"
